### PR TITLE
Add a ton of ELRC data

### PR DESCRIPTION
**Pull request by kpu**
I went overboard and wrote a scraper for ELRC-SHARE: https://github.com/kpu/elrc-scrape/

Could use your help on a few things.  

Some of them require an HTTP post to "agree" to a very open license:
```
wget -O 386.zip --post-data='licence_agree=on&in_licence_agree_form=True&licence=NLOD-1.0' https://elrc-share.eu/repository/download/b97eab8e240d11e7bfe700155d0205027eb084d4e06f4d03a2c392ad8a282f9b/
wget -O 520.zip --post-data='licence_agree=on&in_licence_agree_form=True&licence=NLOD-1.0' https://elrc-share.eu/repository/download/41a00f16e56611e7b7d400155d026706c540eeb489ec4630836e520f118e8921/
wget -O 522.zip --post-data='licence_agree=on&in_licence_agree_form=True&licence=NLOD-1.0' https://elrc-share.eu/repository/download/3be25a38e56c11e7b7d400155d026706f4d9a0c65db441a699b442ef5678e3a4/
wget -O 527.zip --post-data='licence_agree=on&in_licence_agree_form=True&licence=NLOD-1.0' https://elrc-share.eu/repository/download/2c9a951ce57211e7b7d400155d0267061a31ab2e62374612b334a07229e335d9/
wget -O 530.zip --post-data='licence_agree=on&in_licence_agree_form=True&licence=NLOD-1.0' https://elrc-share.eu/repository/download/1d21b0b4e57411e7b7d400155d0267069e02d51abcc947ebaeeef0f13e9e26da/
wget -O 531.zip --post-data='licence_agree=on&in_licence_agree_form=True&licence=NLOD-1.0' https://elrc-share.eu/repository/download/a3107010e57511e7b7d400155d026706e57641e027694f84a21dbae54bd87515/
wget -O 535.zip --post-data='licence_agree=on&in_licence_agree_form=True&licence=NLOD-1.0' https://elrc-share.eu/repository/download/532c2ec2e57a11e7b7d400155d026706ea9134d7106b464ca2225c7240c49ebe/
wget -O 538.zip --post-data='licence_agree=on&in_licence_agree_form=True&licence=NLOD-1.0' https://elrc-share.eu/repository/download/35362e5ae58311e7b7d400155d026706ebdc5fd134aa4e11ad5c5333f7e24cfa/
wget -O 540.zip --post-data='licence_agree=on&in_licence_agree_form=True&licence=NLOD-1.0' https://elrc-share.eu/repository/download/709f316ee58611e7b7d400155d02670611effeefea1b45ff98872bd9149e7bd0/
wget -O 561.zip --post-data='licence_agree=on&in_licence_agree_form=True&licence=NLOD-1.0' https://elrc-share.eu/repository/download/16006a8ce6f511e7b7d400155d0267061a1e5775094d4d0bb5c9e044adbb8386/
wget -O 563.zip --post-data='licence_agree=on&in_licence_agree_form=True&licence=NLOD-1.0' https://elrc-share.eu/repository/download/30a23210e6f711e7b7d400155d026706973f3daf33ce48d6a89b560b8755cc0e/
wget -O 565.zip --post-data='licence_agree=on&in_licence_agree_form=True&licence=NLOD-1.0' https://elrc-share.eu/repository/download/852eb712e6f811e7b7d400155d026706441496bcf9e44efcae6d8d2d31bd7bcc/
```

This one is a zip file inside a zip file :cry: https://elrc-share.eu/repository/download/278f1236443011ea913100155d026706dc6f3bb672cc452c891d08cb9a7bdd69

Non-standard format of https://elrc-share.eu/repository/browse/cef-data-marketplace-multilingual-benchmark-for-the-evaluation-of-cleaning-and-clustering-tools/365a8b821aa011eb913100155d02670611118e05e423402bb729137ecf6ac864 has text files and I suppose we just want the ones labeled parallel.

### CODE REVIEW

**Comment by thammegowda** [comment]
> Some of them require an HTTP post to "agree" to a very open license:

Ah, why couldnt they just use HTTP GET 🤦 
Lets take some time to think about it. 
 `wget` python library is used for GET, which has limited functionality; I am not sure it supports POST. 
  I liked it because it supports progress bar. We can switch to `requests` or such  (but it wont give a nice progress bar). 
🤔  I have already special cased one dataset that did not fit into exisiting schema of Entry. -- JW300  -
https://github.com/thammegowda/mtdata/blob/da506c1f53f5607d9bd6c82ae7ba69338b03b7c0/mtdata/entry.py#L67-L68

Perhaps we can do a similar one for datasets which need POST request.. 


> This one is a zip file inside a zip file 😢 https://elrc-share.eu/repository/download/278f1236443011ea913100155d026706dc6f3bb672cc452c891d08cb9a7bdd69

We will have to modify file parsers. zip inside zip -- a recursive zip based on extension in `in_paths`   
luckily, `zip` in ISO 639-3 is unused so no confusions there for .zip (but we cant do the same for tar and tgz)
``` mtdata-iso zip tar tgz       
Input   ISO639_3        Name
zip     -none-  -none-
tar     tar     Central Tarahumara
tgz     tgz     Tagalaka
```

**Comment by kpu** [comment]
I've written ELRC to try and convince them that NLOD-1.0 shouldn't require an HTTP post.

**Comment by kpu** [comment]
And as regards the zip file in a zip file, maybe it is easier to get it rehosted on e.g. OPUS or even statmt.

**Comment by kpu** [comment]
I've tried to name everything semi-automatically though periods in domains #50 would be nice.

**Comment by thammegowda** [comment]
I am wondering if we can simplify this file by separating code from data. 
i.e.  create TSV or JSON file  then code can be simple as this: 
```
for row in <parse(data)>
  index.add_entry(Entry(**row))
```
This will reduce the burden of semi-automatically generating python code. 
also, alphabetically sorting rows in data, will help make future modifications (add/remove/update) easier (especially useful for viewing git diffs). 

FYI, it is possible to include data files in python package (say JSON or TSV or such) by listing it in MANIFEST.in, similar to this https://github.com/thammegowda/mtdata/blob/47d2aac74fc40c0621b8e21483b298a274cf57fb/MANIFEST.in#L2 

and read it using relative path e.g. https://github.com/thammegowda/mtdata/blob/47d2aac74fc40c0621b8e21483b298a274cf57fb/mtdata/index/__init__.py#L11

**Comment by kpu** [comment]
With weird corpora banned and the post-requiring corpora filtered out, it's time to test how many TMXes fail to parse.  Then how many have text in the languages they claim to be in (I've caught a few in the wrong language).

**Comment by kpu** [comment]
@thammegowda I have an explicit list of TMXes for each language pair but ran into
```
  File "mtdata/mtdata/parser.py", line 42, in __post_init__
    assert 1 <= len(self.paths) <= 3
AssertionError
```

What's the right way to do this?  

Some corpora have multiple TMX files for no clear reason and they should just be concatenated:
https://elrc-share.eu/repository/download/3874ddfe6d1b11e7b7d400155d026706aa408cdca7fa41c681819c8ad1d3efd2/  For this *.tmx would work though it's a special case.  

The general case has multiple files per language pair and multiple language pairs: https://elrc-share.eu/repository/browse/memorias-de-traduccion-portal-oficial-de-turismo-de-espana-wwwspaininfo/04dfcaec9ca011e9a7e100155d02670640ce598203e246e8ae4fe838be231213/

**Comment by thammegowda** [comment]
That assertion guards the assumptions made in this part of code:
https://github.com/thammegowda/mtdata/blob/47d2aac74fc40c0621b8e21483b298a274cf57fb/mtdata/parser.py#L79-L85 

-  one path => assume reader produces (source, target) from one file
-  two paths  => assume source and target are in separate paths, we read in parallel
- I am not entirely sure why I added `len(self.paths) <= 3`  instead of `len(self.paths) <= 2` -- I think it is for OPUS JW300 where source, target and alignment are in three separate files.. 

This is the only concern for removing/broadening that assertion. 
I think it is fair to assume if all paths are `tmx` , then we can read paths sequentially and yield all records.

**Comment by thammegowda** [comment]
BTW, the TSV file looks good! thanks for doing that.

**Comment by kpu** [comment]
Brute-force checking all the files.  
```
mtdata list |grep ELRC |while read a; do l=$(cut -f 2 <<<"$a"); d=$(cut -f 1 <<<"$a"); dir=$HOME/data/test/$l/$d; mkdir -p $dir; echo mtdata get -tr $d -l $l -o $dir \>$dir/out 2\>$dir/err \&\& touch $dir/success; done |parallel -j4
```

Here by the way is a list of issues I've sent to ELRC:
https://docs.google.com/spreadsheets/d/1NrvmSadH0nABAyDYBzznt8hreVyf5IqaAqvFb4qKtmE/edit?usp=sharing

**Comment by thammegowda** [comment]
Wow, this is indeed a lot of data. Kudos to you for getting to the bottom of this issue 🎩 
The PR looks so good to me and I am eager to merge.

File: mtdata/parser.py [hunk 1]
```diff
@@ -78,6 +80,9 @@ def read_segs(self):
 
         if len(readers) == 1:
             yield from readers[0]
+        elif self.ext == 'tmx' or self.ext == 'tsv':
+            for reader in readers:
+                yield from reader
```
Comment (thammegowda) [inline mtdata/parser.py:85]: 💯

**Comment by kpu** [comment]
Think I've banned everything that doesn't parse or is grossly in the wrong language.  Doing another pass at extracting everything.  

Keep in mind the NTEU data is actually a compilation of other corpora as per the ReadMe.txt in each zip.  And that's what led to adding EU corpora.  I think the only part mtdata is missing is DCEP #40.

**Comment by kpu** [comment]
Pass completed.  Presuming they haven't updated files, everything currently parses correctly.  I say we put it in then do an update when ELRC fixes things.  They've started fixing some already.

**Comment by thammegowda** [comment]
I randomly picked some datasets and tested them. All my tests passed.

But one small thing I noticed.

For example, these two come from dame URL/zip and that zip downloaded only once. 
```
python -m mtdata get -l eng-spa -tr ELRC_Praias_2007 -o tmp/elrc_praias
python -m mtdata get -l deu-por -tr ELRC_Praias_2007 -o tmp/elrc_praias
```

Similarly these TierA  and TierB are also from a single zip
```
python -m mtdata get -l dan-hrv -tr ELRC_NTEU_TierA ELRC_NTEU_TierB -o tmp/elrc
# https://elrc-share.eu/repository/download/ab977a34b3fe11eb9c1a00155d026706fb08f03bfca844098fc1fe93a465a3f0/
```

But that zip gets downloaded two times since they are assigned with different `filename`s each time.
- $HOME/.mtdata/elrc-share.eu/1b2a/a8ce22ce9c2f61140996310ca888/ELRC_NTEU_TierA.zip
- $HOME/.mtdata/elrc-share.eu/1b2a/a8ce22ce9c2f61140996310ca888/ELRC_NTEU_TierB.zip

So a little bit of inefficiency.
I think this okay for this source for now, given how chaotic their formats are and we might be updating them in near future.

**Comment by thammegowda** [comment]
@kpu Merged!  Thank you very much.

**Pull request merged**